include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+      // will end with a single same T object stored in each element of the vector:
+      typename std::map<CellIteratorType,CellVector<DataType> >::iterator it = map.find(cell);
+      for (unsigned int q=0; q < n_q_points; q++)
+        it->second[q] = std::make_shared<T>();
+    }
+}
+
+template <typename CellIteratorType, typename DataType>
+template<typename T>
+void CellDataStorage<CellIteratorType,DataType>::initialize(const CellIteratorType &cell_start,
+                                                            const CellIteratorType &cell_end,
+                                                            const unsigned int number)
+{
+  for (CellIteratorType it = cell_start; it != cell_end; it++)
+    if (it->is_locally_owned())
+      initialize<T>(it,number);
+}
+
+
+
+template <typename CellIteratorType, typename DataType>
+bool CellDataStorage<CellIteratorType,DataType>::erase(const CellIteratorType &cell)
+{
+  const typename std::map<CellIteratorType,CellVector<DataType> >::const_iterator it = map.find(cell);
+  for (unsigned int i = 0; i < it->second.size(); i++)
+    {
+      Assert(it->second[i].unique(),
+             ExcMessage("Can not erase the cell data multiple objects reference its data."));
+    }
+
+  return (map.erase(cell)==1);
+}
+
+template <typename CellIteratorType, typename DataType>
+void CellDataStorage<CellIteratorType,DataType>::clear()
+{
+  // Do not call
+  // map.clear();
+  // as we want to be sure no one uses the stored objects. Loop manually:
+  typename std::map<CellIteratorType,CellVector<DataType> >::iterator it = map.begin();
+  while (it != map.end())
+    {
+      // loop over all objects and see if noone is using them
+      for (unsigned int i = 0; i < it->second.size(); i++)
+        {
+          Assert(it->second[i].unique(),
+                 ExcMessage("Can not erase the cell data multiple objects reference its data."));
+        }
+      it = map.erase(it);
+    }
+}
+
+template <typename CellIteratorType, typename DataType>
+template <typename T>
+CellDataStorage<CellIteratorType,DataType>::CellVector<T>
+CellDataStorage<CellIteratorType,DataType>::get_data(const CellIteratorType &cell)
+{
+  static_assert(std::is_base_of<DataType, T>::value,
+                "User's T class should be derived from user's DataType class");
+
+  typename std::map<CellIteratorType,CellVector<DataType> >::iterator it = map.find(cell);
+  Assert(it != map.end(), ExcMessage("Could not find data for the cell"));
+
+  // It would be nice to have a specilized version of this function for T==DataType.
+  // However explicit (i.e full) specialization of a member template is only allowed
+  // when the enclosing class is also explicitly (i.e fully) specialized.
+  // Thus, stick with copying of shared pointers even when the T==DataType:
+  CellVector<T> res (it->second.size());
+  for (unsigned int q = 0; q < res.size(); q++)
+    res[q] = std::dynamic_pointer_cast<T>(it->second[q]);
```

Comment:
I think you can avoid all of this by saying
```
  CellVector<T> res (it->second.begin(), it->second.end());
  return res;
```
This will automatically do the casts (if necessary) of the objects.

In fact, this should work as well:
```
  return CellVector<T> (it->second.begin(), it->second.end());
```